📝 Add Documentation (#26)
pcaversaccio/createx: README.md

  - [Special Features](#special-features)
    - [Permissioned Deploy Protection and Cross-Chain Redeploy Protection](#permissioned-deploy-protection-and-cross-chain-redeploy-protection)
    - [Pseudo-Random Salt Value](#pseudo-random-salt-value)
  - [Design Principles](#design-principles)
  - [Security Considerations](#security-considerations)
  - [Tests](#tests)
    - [Test Coverage](#test-coverage)
  - [ABI (Application Binary Interface)](#abi-application-binary-interface)
  - [New Deployment(s)](#new-deployments)
    - [Contract Verification](#contract-verification)
  - [`CreateX` Deployments](#createx-deployments)

## Available Versatile Functions

```ml
CreateX
├── CREATE
│   ├── Read-Only Functions
│   │   ├── "computeCreateAddress(uint256) view returns (address)"
│   │   └── "computeCreateAddress(address,uint256) view returns (address)"
│   └── Write Functions
│       ├── "function deployCreate(bytes) payable returns (address)"
│       ├── "function deployCreateAndInit(bytes,bytes,tuple(uint256,uint256)) payable returns (address)"
│       ├── "function deployCreateAndInit(bytes,bytes,tuple(uint256,uint256),address) payable returns (address)"

> │       └── "deployCreateClone(address,bytes) payable returns (address)"

Author: Great catch! Fixed via https://github.com/pcaversaccio/createx/commit/eb81a63af947fdf41ed8ec419741bfa29e19f31e.

  ) {
    // Configures solely a permissioned deploy protection.
    guardedSalt = _efficientHash({
      a: bytes32(uint256(uint160(msg.sender))),
      b: salt
    });
  } else if (senderBytes == SenderBytes.MsgSender) {
    // Reverts if the 21st byte is greater than `0x01` in order to enforce developer explicitness.
    revert InvalidSalt({ emitter: _SELF });
  } else if (
    senderBytes == SenderBytes.ZeroAddress &&
    redeployProtectionFlag == RedeployProtectionFlag.True
  ) {
    // Configures solely a cross-chain redeploy protection. In order to prevent a pseudo-randomly
    // generated cross-chain redeploy protection, we enforce the zero address check for the first 20 bytes.
    guardedSalt = _efficientHash({ a: bytes32(block.chainid), b: salt });
  } else if (
    senderBytes == SenderBytes.ZeroAddress &&
    redeployProtectionFlag == RedeployProtectionFlag.Unspecified
  ) {
    // Reverts if the 21st byte is greater than `0x01` in order to enforce developer explicitness.
    revert InvalidSalt({ emitter: _SELF });
  } else {
    // In all other cases, the salt value `salt` is not modified.
    guardedSalt = salt;
  }
}
```

> Please note that when you configure a permissioned deploy protection, you **must** specify whether you want a cross-chain redeploy protection or not (i.e. the 21st byte equals `0x00`). The underlying reason for this logic is to enforce developer explicitness. If you don't configure a cross-chain redeploy protection, i.e. the 21st byte is greater than `0x01`, the function reverts. Furthermore, you can configure solely a cross-chain redeploy protection by setting the first 20 bytes equal to the zero address `0x0000000000000000000000000000000000000000`. The rationale behind this logic is to prevent a pseudo-randomly generated cross-chain redeploy protection. Also in this case, if you don't specify a cross-chain redeploy protection, i.e. the 21st byte is greater than `0x01`, the function reverts. The underlying reason for this logic is as well to enforce developer explicitness.

Reviewer: Nit on phrasing, I think this might be a more clear
```suggestion
Please note that when you configure a permissioned deploy protection, you **must** specify whether you want cross-chain redeploy protection (i.e. 21st byte equals `0x01`) or not (i.e. the 21st byte equals `0x00`). The underlying reason for this logic is to enforce developer explicitness. If you don't specify a cross-chain redeploy protection decision (i.e. the 21st byte is greater than `0x01`) the function reverts.

Furthermore, you can configure _only_ cross-chain redeploy protection by setting the first 20 bytes equal to the zero address `0x0000000000000000000000000000000000000000`. The rationale behind this logic is to prevent a pseudo-randomly generated 32 byte salt from inadvertently activating cross-chain redeploy protection. Also in this case, if you don't specify a cross-chain redeploy protection, i.e. the 21st byte is greater than `0x01`, the function reverts. The underlying reason for this logic is as well to enforce developer explicitness.
```

### Pseudo-Random Salt Value

For developer convenience, the [`CreateX`](./src/CreateX.sol) contract offers several overloaded functions that generate the salt value pseudo-randomly using a diverse selection of block and transaction properties. Please note that this approach does not guarantee true randomness!

The full logic is implemented in the `internal` [`_generateSalt`](./src/CreateX.sol#L965-L991) function:

```solidity
function _generateSalt() internal view returns (bytes32 salt) {

>   // If you use this function between the genesis block and block number 31, it will return zero.

Reviewer: Ah this is still a bit misleading actually. In that block range, the value of `blockhash(block.number - 32)` is zero, but the resulting salt from the entire hash is still some nonzero value based on the remaining properties

Author: Great callout! Adjusted via https://github.com/pcaversaccio/createx/commit/1dcdf248121d2d503b5ac4aad4db08c1a77b3a03.

  unchecked {
    salt = keccak256(
      abi.encode(
        // We don't use `block.number - 256` (the maximum value on the EVM) to accommodate
        // any chains that may try to reduce the amount of available historical block hashes.
        // We also don't subtract 1 to mitigate any risks arising from consecutive block
        // producers on a PoS chain. Therefore, we use `block.number - 32` as a reasonable
        // compromise, one we expect should work on most chains, which is 1 epoch on Ethereum
        // mainnet.
        blockhash(block.number - 32),
        block.coinbase,
        block.number,
        block.timestamp,
        block.prevrandao,
        block.chainid,
        msg.sender
      )
    );
  }
}
```

## Design Principles

- [`CreateX`](./src/CreateX.sol) should cover _most_ but not all contract creation use cases.
- [`CreateX`](./src/CreateX.sol) should be human-readable and should be simple to understand for readers with low prior experience.
- [`CreateX`](./src/CreateX.sol) should be maximally secure, i.e. no hidden footguns.
- [`CreateX`](./src/CreateX.sol) should be trustless.
- [`CreateX`](./src/CreateX.sol) should be stateless.

> - [`CreateX`](./src/CreateX.sol) should be extensible.

Reviewer: We say "extensible" a few times but I think it might not be clear what we have in mind by that
```suggestion
- [`CreateX`](./src/CreateX.sol) should be extensible (i.e. it can be used to deploy protocols, within protocols, or to deploy other types of deterministic deployer factories).
```

- The functions [`deployCreate3(bytes)`](https://github.com/pcaversaccio/createx/blob/main/src/CreateX.sol#L648-L665), [`deployCreate3AndInit(bytes,bytes,tuple(uint256,uint256))`](https://github.com/pcaversaccio/createx/blob/main/src/CreateX.sol#L801-L834), and [`deployCreate3AndInit(bytes,bytes,tuple(uint256,uint256),address)`](https://github.com/pcaversaccio/createx/blob/main/src/CreateX.sol#L764-L799) do not implement any permissioned deploy protection, thus anyone can frontrun the same proxy deployment on other chains. Use with caution!

## Tests

> For all (fuzz) tests available in the [`test`](./test) directory, we have consistently applied the [Branching Tree Technique](https://twitter.com/PaulRBerg/status/1682346315806539776). This means that each test file is accompanied by a `.tree` file that defines all the necessary branches to be tested.

Reviewer: ```suggestion
For all (fuzz) tests available in the [`test`](./test) directory, we have consistently applied the [Branching Tree Technique](https://twitter.com/PaulRBerg/status/1682346315806539776) with [bulloak](https://github.com/alexfertel/bulloak). This means that each test file is accompanied by a `.tree` file that defines all the necessary branches to be tested.
```

Author: Fixed via https://github.com/pcaversaccio/createx/pull/26/commits/8a1783bcf2719c3d0fb0d5bf0d14200729c323a1. Wanted to use code highlighting for `bulloak` & made you a co-author of the commit of course.

| src/CreateX.sol | 100.00% (149/149) | 100.00% (210/210) | 100.00% (78/78) | 100.00% (31/31) |
```

> **Note**<br>
> A test coverage of 100% does not mean that there are no vulnerabilities. What really counts is the quality and spectrum of the tests themselves!

## ABI (Application Binary Interface)

<details>

> <summary> <a href="https://docs.soliditylang.org/en/latest/">Solidity</a> </summary>

Reviewer: Should we instead link to the `ICreateX.sol` file to make that more discoverable, and users can then `forge install` this repo?

Author: I think we can do both. Keeping it in the `README` (I like the easy copy+paste feature from GH for these cases) and add a link. Added a note here with the link: https://github.com/pcaversaccio/createx/pull/26/commits/9ee14ae3bfd6322602d8adac01602c6fa20febfc. Also, it's already on the top of the `README` 
![image](https://github.com/pcaversaccio/createx/assets/25297591/8d57fe65-0578-4fb1-9773-1bd2fdcbc65a)

  "function deployCreate2Clone(address implementation, bytes data) payable returns (address proxy)",
  "function deployCreate3(bytes initCode) payable returns (address newContract)",
  "function deployCreate3(bytes32 salt, bytes initCode) payable returns (address newContract)",
  "function deployCreate3AndInit(bytes32 salt, bytes initCode, bytes data, (uint256 constructorAmount, uint256 initCallAmount) values) payable returns (address newContract)",
  "function deployCreate3AndInit(bytes initCode, bytes data, (uint256 constructorAmount, uint256 initCallAmount) values) payable returns (address newContract)",
  "function deployCreate3AndInit(bytes32 salt, bytes initCode, bytes data, (uint256 constructorAmount, uint256 initCallAmount) values, address refundAddress) payable returns (address newContract)",
  "function deployCreate3AndInit(bytes initCode, bytes data, (uint256 constructorAmount, uint256 initCallAmount) values, address refundAddress) payable returns (address newContract)",
  "function deployCreateAndInit(bytes initCode, bytes data, (uint256 constructorAmount, uint256 initCallAmount) values) payable returns (address newContract)",

>   "function deployCreateAndInit(bytes initCode, bytes data, (uint256 constructorAmount, uint256 initCallAmount) values, address refundAddress) payable returns (address newContract)",

Reviewer: Interesting, viem ABIs can be defined with structs, e.g.
```javascript
"struct Values { uint256 constructorAmount; uint256 initCallAmount; }",
"function deployCreateAndInit(bytes initCode, bytes data, Values memory values) payable returns (address newContract)",
"function deployCreateAndInit(bytes initCode, bytes data, Values memory values, address refundAddress) payable returns (address newContract)",
```

so I wonder why it didn't come out this way.  I forget how I generated [this `Multicall3` one](https://github.com/mds1/multicall3-frontend/blob/7df0684d01cafa57cc729329f35c1b6ce2a9d8b5/src/lib/constants.ts#L246-L267) to get it in that format. Not a big issue though

Author: I used your script [here](https://[github.com/mds1/evm-diff/blob/main/script/json-abi-to-viem-human-readable.ts](https://github.com/mds1/evm-diff/blob/main/script/json-abi-to-viem-human-readable.ts). In your case, the struct forms an array, maybe because of this you generated the struct? Can you please confirm that I generated the correct viem ABI using your script. Otherwise, paste it as suggested change and I will commit.

Author: Interestingly, the viem guys also imply [here](https://github.com/wagmi-dev/viem/issues/1406#issuecomment-1783916066) to use it as I generated it; so I guess it's fine:

```md
That is an Ethers flavored ABI. Viem only supports generic flavored ABIs. To fix this, simply find and replace the word ”tuple” with ””.
```

Reviewer: Agreed this is fine 👌

TBD (section on pre-signed transactions; we will offer multiple pre-signed transactions with different `gasLimit` levels; will be added after the feedback phase)

### Contract Verification

To verify a deployed [`CreateX`](./src/CreateX.sol) contract on a block explorer, use the following parameters:

- TBD (will be added after the feedback phase)

## [`CreateX`](./src/CreateX.sol) Deployments

> **Warning**<br>
> The address `0x0000000000000000000000000000000000000000` is a simple placeholder for now. Do not send any funds there!

- EVM-Based Production Networks:
  - Ethereum: [`0x0000000000000000000000000000000000000000`](https://etherscan.io/address/0x0000000000000000000000000000000000000000)
- Ethereum Test Networks:
  - Goerli: [`0x0000000000000000000000000000000000000000`](https://goerli.etherscan.io/address/0x0000000000000000000000000000000000000000)
- Additional EVM-Based Test Networks:

>   - Optimism Testnet (Goerli): [`0x0000000000000000000000000000000000000000`](https://goerli-optimism.etherscan.io/address/0x0000000000000000000000000000000000000000)

Reviewer: What do you think about the [JSON file](https://github.com/mds1/multicall/blob/main/deployments.json) approach I used for Multicall3 instead? We can fork the https://www.multicall3.com/ frontend too, that way it's easy to have a table searchable by name or Chain ID. Can also add any other fields/info you want to the JSON and table in the UI if we go that route

(Side benefit is some extra promotion when people [tweet](https://twitter.com/MaximilianFSN/status/1719391868146909191)) out their chain on the website 😅

Author: That's actually only a placeholder 😄 - I personally like having everything in one place (in our case in the `README`), but we can the `JSON` file approach as well mirror it later in a UI & add some further infos like `chainId`, link to e.g. EVM Diff etc. So IMO, the `README` should still have it, but without all the additional information. Would you agree with this approach? Also, we can decide on this in 2-3 weeks after the feedback phase.

Reviewer: Cool, that sounds good to me! If we stand up a frontend I can manage that part
